### PR TITLE
Refactor Nested Rules

### LINKED ISSUE #55: Nested rules are ignored if both included and excluded
- [ ] Fix scenario where groupSets does not get included in the metadata package.

- [ ] Use the _.path utility to parse the nestedRules.

### PR DESCRIPTION
**Pull request by SferaDev**
### :pushpin: References

* **Issue:** Closes #55 

### :memo: Implementation

- Use lodash to parse nested rules.

### :fire: Is there anything the reviewer should know to test it?

- GroupSets are now properly included in the metadata package.

### CODE REVIEW

**Comment by adrianq** [comment]
@SferaDev This is ready to be tested by @tokland and myself?

**Comment by SferaDev** [comment]
It works fine but I was still revising if it's good enough.

I'll do some final tests this afternoon while I finish other tasks and add you as reviewers when sure.

File: src/utils/synchronization.ts [hunk 1]
```diff
@@ -5,21 +5,29 @@ import "../utils/lodash-mixins";
...
-        {}
-    );
+export function buildNestedRules(rules: string[][] = []): NestedRules {
+    return _(rules)
+        .filter(path => path.length > 1)
+        .groupBy(_.head)
+        .tap(object => {
+            _.forOwn(object, (array, key) => (object[key] = array.map(rule => rule.slice(1))));
+        })
+        .value();
```
Comment (tokland) [inline src/utils/synchronization.ts]: _.mapValues comes in handy:

```
return _(rules)
        .filter(path => path.length > 1)
        .groupBy(_.head)
        .mapValues(path => path.map(_.tail))
        .value();
```

File: src/utils/synchronization.ts [hunk 2]
```diff
@@ -5,21 +5,29 @@ import "../utils/lodash-mixins";
...
+export function cleanObject(element: any, excludeRules: string[][] = []): any {
+    const leafRules = excludeRules.filter(array => array.length === 1).map(array => array[0]);
+    return _.pick(element, _.difference(_.keys(element), leafRules));
+}
+
+export function cleanReferences(
+    references: MetadataPackage,
+    includeRules: string[][] = []
+): string[] {
+    return _.intersection(_.keys(references), includeRules.map(array => array[0]));
```
Comment (tokland) [inline src/utils/synchronization.ts]: Very Minor: A bit more compact `rules.map(_.first)`

**Comment by SferaDev** [comment]
Yes, we will need to focus on adding a battery of tests at the start of the second milestone.

Thanks, will update the PR with your comments in about an hour.

**Comment by adrianq** [comment]
Thanks @SferaDev I have deeply tested this PR and it seems to be working fine. As discussed, we will have to set up some tests before implementing the other sync features.

**Pull request merged**